### PR TITLE
Add 'hidden' state to dkan_publishing workflow

### PR DESCRIPTION
**Pull request by tiffneybare**
fixes WCMS-4889

Adding logic that looks at the moderation state and if the moderation state equals publish_without_search_indexing not to index the dataset in the dkan search API. 

## QA Steps

1. Edit an existing dataset or create a new dataset.
  a. Select the 'Publish without search indexing' moderation state at the bottom of the form.
  b. Click Save.
2. Search for the dataset you have 'hidden'.
  a. The dataset should not show up in the search results.
3. Navigate to the hidden dataset.
  a. It should be visible.
4. Unhide & publish the hidden dataset.
5. Search for the dataset.
  a. It should show up in the search results.

### CODE REVIEW

**Comment by clayliddell** [comment]
@dafeder Are the upgrade tests not importing config changes? That's the only reason I can think of why they might be failing...

Update: Response from @dafeder on Slack:

> @clayliddell upgrade tests just run updatedb. If config change you need to write an update hook

File: modules/metastore/modules/metastore_search/metastore_search.module
Comment (dafeder) [inline modules/metastore/modules/metastore_search/metastore_search.module]: @clayliddell isn't the third condition here all we need? We could wrap the `new Data()` call in a try/catch.

File: modules/metastore/modules/metastore_search/metastore_search.module
Comment (dafeder) [inline modules/metastore/modules/metastore_search/metastore_search.module]: I think it's best not to have a function name this generic in the global namespace... we could easily end up defining a function with the same name in any other dkan module. can we rename to `_dkan_metastore_search_is_dataset()`?

File: modules/metastore/src/Controller/MetastoreController.php
Comment (dafeder) [inline modules/metastore/src/Controller/MetastoreController.php]: Not clear on why this is necessary; the [service get method](https://github.com/GetDKAN/dkan/blob/387959f740b9734fcc01808d612ba16b5591042a/modules/metastore/src/Service.php#L241) will only [retrieve published datasets](https://github.com/GetDKAN/dkan/blob/387959f740b9734fcc01808d612ba16b5591042a/modules/metastore/src/Storage/Data.php#L187). 

If it is necessary, I don't think `InvalidArgumentException` is appropriate here. The arguments are valid. You could use `Drupal\metastore\Exception\MissingObjectException`.

**Review by dafeder — CHANGES_REQUESTED** [review]
Couple questions in the comments -- looks very good but I'd like to check in and make sure I understand the plugin/processor architecture here and why it was necessary.

File: modules/metastore/modules/metastore_search/metastore_search.module
Comment (clayliddell) [reply modules/metastore/modules/metastore_search/metastore_search.module]: Yeah, I agree. It's a bit less readable, but it at least guarantees consolidation of validation logic.

File: modules/metastore/modules/metastore_search/metastore_search.module
Comment (clayliddell) [reply modules/metastore/modules/metastore_search/metastore_search.module]: Fair enough.

File: modules/metastore/src/Controller/MetastoreController.php
Comment (clayliddell) [reply modules/metastore/src/Controller/MetastoreController.php]: I think you're right. This was an early change that was made which can be removed due to later refactoring.

**Review by dafeder — APPROVED** [review]

File: modules/metastore/config/install/workflows.workflow.dkan_publishing.yml
Comment (dafeder) [inline modules/metastore/config/install/workflows.workflow.dkan_publishing.yml]:
```suggestion
      label: 'Published (hidden)'
```
More consistent state labels

File: modules/metastore/metastore.install
Comment (dafeder) [inline modules/metastore/metastore.install]:
```suggestion
    'label' => 'Published (hidden)',
```
More consistent state labels

**Review by dafeder — CHANGES_REQUESTED** [review]
Sorry for a second round, this came up after more careful QA. I think the naming needs to be more consistent with the other states, and I think "search indexing" is a confusing term for a non-Drupal-expert. Any name is still potentially confusing though and we should figure out a way to inject more contextual help in the form later on.

**Comment by dafeder** [comment]
Applied suggestions after 👍🏽 from @clayliddell , will merge if tests still pass.

**Review by dafeder — APPROVED** [review]

**Pull request merged**
